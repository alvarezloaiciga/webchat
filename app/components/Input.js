// @flow
import * as React from 'react';
import debounce from 'lodash/debounce';
import messages from 'core-ui/messages';
import type {IntlMessage} from 'types';
import {getDisplayString} from 'core-ui/services/i18nService';

export type InputProps = {
  value?: string | null, // Null means parent doesn't want to track value
  initialValue?: string,
  onClick?: () => void,
  onBlur?: () => void,
  onFocus?: () => void,
  onChange?: (e: SyntheticInputEvent<>) => void,
  onKeyDown?: (e: SyntheticKeyboardEvent<> & SyntheticInputEvent<>) => void,
  onMouseOver?: () => void,
  onMouseOut?: () => void,
  onSubmit?: (value: string) => void,
  required?: boolean, // Presentational
  disabled?: boolean,
  maxLength?: number,
  debounce?: number,
  className?: string,
  inputStyle?: string | Object,
  placeholder?: IntlMessage | string,
  label?: IntlMessage | string,
  title?: IntlMessage | string,
  autoFocus?: boolean,
  'data-test'?: string, //eslint-disable-line react/no-unused-prop-types
};

type InputState = {
  value: string,
};

export class Input extends React.Component<InputProps, InputState> {
  state: InputState = {
    value: '',
  };
  static defaultProps = {
    value: '',
  };
  debouncedOnChange: any;
  input: any;

  componentDidMount() {
    /* eslint-disable react/no-did-mount-set-state */
    this.setState({value: this.props.initialValue || this.props.value || ''}, () => {
      if (this.props.onChange) {
        /*
         * We require this intermediate function because react discards the synthetic event
         * with the debounce call. Therefore without this function to persist it,
         * target would be undefined.
         */
        this.debouncedOnChange = this.props.debounce
          ? debounce(this.props.onChange, this.props.debounce)
          : this.props.onChange;
      }
    });
  }

  componentWillReceiveProps(nextProps: InputProps) {
    if (
      nextProps.value !== undefined &&
      nextProps.value !== null &&
      (nextProps.value !== this.props.value || nextProps.value !== this.state.value)
    ) {
      this.setState({value: nextProps.value});
    }
  }

  onChange = (e: SyntheticInputEvent<>) => {
    const value = e.target.value;
    e.persist();
    this.setState({value}, () => {
      if (this.props.onChange) {
        this.debouncedOnChange(e);
      }
    });
  };

  onKeyDown = (e: SyntheticKeyboardEvent<> & SyntheticInputEvent<>) => {
    if (e.keyCode === 13 && this.props.onSubmit) {
      this.props.onSubmit(this.state.value);
      e.preventDefault();
    } else if (this.props.onKeyDown) {
      this.props.onKeyDown(e);
    }
  };

  getValue = () => this.state.value;

  /* eslint-disable jsx-a11y/no-autofocus */
  render() {
    return (
      <div className={`Input ${this.props.className || ''}`}>
        {this.props.label && (
          <label className="inputLabel" title={getDisplayString(this.props.title)}>
            {getDisplayString(this.props.label)}
            {this.props.required && (
              <span
                className="required"
                onMouseOver={this.props.onMouseOver}
                onMouseOut={this.props.onMouseOut}
                title={getDisplayString(messages.required)}
              >
                {' '}
                *
              </span>
            )}
          </label>
        )}
        <input
          ref={r => {
            this.input = r;
          }}
          className={typeof this.props.inputStyle === 'string' ? this.props.inputStyle : ''}
          type="text"
          title={getDisplayString(this.props.title)}
          placeholder={getDisplayString(this.props.placeholder)}
          onClick={this.props.onClick}
          onChange={this.onChange}
          maxLength={this.props.maxLength}
          onKeyDown={this.onKeyDown}
          onBlur={this.props.onBlur}
          onFocus={this.props.onFocus}
          autoFocus={this.props.autoFocus}
          data-test={this.props['data-test']}
          onMouseOver={this.props.onMouseOver}
          onMouseOut={this.props.onMouseOut}
          value={this.state.value}
          style={typeof this.props.inputStyle === 'object' ? this.props.inputStyle : undefined}
          disabled={this.props.disabled}
        />
      </div>
    );
  }
}

export default Input;
