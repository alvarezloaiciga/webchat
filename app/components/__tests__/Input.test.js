// @flow
import * as React from 'react';
import type {InputProps} from '../Input';
import Input from '../Input';
import {shallow} from 'enzyme';
import type {ShallowWrapper} from 'enzyme';

describe('Input component', () => {
  let wrapper: ShallowWrapper;
  let testProps: InputProps;
  let render: () => void;
  let input: ShallowWrapper;
  let instance: any;

  const simulateInput = (value?: string) => {
    input.simulate('change', {target: {value}, persist: jest.fn()});
  };

  beforeEach(() => {
    testProps = {
      value: 'hi',
      onChange: jest.fn(),
      onKeyDown: jest.fn(),
      required: false,
      disabled: false,
      debounce: undefined,
      className: 'crazzzyyyClass',
      label: 'crazzzzyyyLabel',
      title: 'crazzzzyyyTitle',
      placeholder: 'crazzzzyyyPlaceholder',
    };
    render = () => {
      wrapper = shallow(<Input {...testProps} />);
      instance = wrapper.instance();
      instance.componentDidMount();
      input = wrapper.find('input').at(0);
    };
  });

  describe('rendering', () => {
    it('renders with all props', () => {
      render();
      expect(wrapper).toMatchSnapshot();
    });

    it('renders with only required props', () => {
      testProps = {onChange: jest.fn()};
      render();
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('onChange', () => {
    beforeEach(() => {
      render();
    });

    describe('when required is false', () => {
      describe('when input is blank', () => {
        it('calls onChange', () => {
          simulateInput('');
          expect(testProps.onChange).toBeCalledWith(expect.objectContaining({target: {value: ''}}));
        });
      });

      describe('when input is not blank', () => {
        it('calls onChange', () => {
          simulateInput('newStuff');
          expect(testProps.onChange).toBeCalledWith(
            expect.objectContaining({target: {value: 'newStuff'}}),
          );
        });
      });
    });
  });

  describe('componentWillReceiveProps', () => {
    beforeEach(() => {
      render();
    });

    describe('when value is a string', () => {
      it('updates state', () => {
        render();

        testProps.value = 'newValue';
        instance.componentWillReceiveProps(testProps);
        expect(wrapper.state('value')).toBe('newValue');
      });
    });

    describe('when value is undefined', () => {
      it('does not update state', () => {
        render();

        const oldValue = testProps.value;
        testProps.value = undefined;
        instance.componentWillReceiveProps(testProps);
        expect(wrapper.state('value')).toBe(oldValue);
      });
    });

    describe("when value didn't change", () => {
      it('does not updates state', () => {
        render();
        instance.componentWillReceiveProps(testProps);
        expect(wrapper.state('value')).toBe(testProps.value);
      });
    });
  });
});
