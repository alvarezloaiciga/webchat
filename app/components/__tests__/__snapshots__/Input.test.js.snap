// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component rendering renders with all props 1`] = `
<div
  className="Input crazzzyyyClass"
>
  <label
    className="inputLabel"
    title="crazzzzyyyTitle"
  >
    crazzzzyyyLabel
  </label>
  <input
    className=""
    disabled={false}
    onChange={[Function]}
    onKeyDown={[Function]}
    placeholder="crazzzzyyyPlaceholder"
    title="crazzzzyyyTitle"
    type="text"
    value="hi"
  />
</div>
`;

exports[`Input component rendering renders with only required props 1`] = `
<div
  className="Input "
>
  <input
    className=""
    onChange={[Function]}
    onKeyDown={[Function]}
    placeholder=""
    title=""
    type="text"
    value=""
  />
</div>
`;
